### PR TITLE
NP-46971-swap-published-file-for-accepted-file-for-academic-article

### PR DESCRIPTION
**Pull request by anetteOlli**

### CODE REVIEW

File: brage-import/src/main/java/no/sikt/nva/brage/migration/merger/AssociatedArtifactComparator.java [hunk 1]
```diff
@@ -1,46 +0,0 @@
-package no.sikt.nva.brage.migration.merger;
-
-import java.util.List;
```
Comment (anetteOlli) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/merger/AssociatedArtifactComparator.java:3]: Removed unused class

File: brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java [hunk 2]
```diff
@@ -166,33 +172,61 @@ private Set<AdditionalIdentifier> mergeAdditionalIdentifiers() {
...
+
+    private boolean academicArticleRulesApply() {
+        return isAcademicArticle()
+               && noneOfTheExistingFilesArePublishedVersion(extractPublishedFiles(cristinPublication))
+               && brageFileIsPublishedVersion(extractPublishedFiles(bragePublicationRepresentation.publication()));
     }
 
-    private boolean shouldUseBrageArtifacts() {
-        return bragePublicationHasAssociatedArtifacts()
-               && (cristinPublication.getAssociatedArtifacts().isEmpty()
```
Comment (anetteOlli) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java:194]: The isEmpty check is redundant, it has already been checked at line 177.

File: brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java [hunk 3]
```diff
@@ -166,33 +172,61 @@ private Set<AdditionalIdentifier> mergeAdditionalIdentifiers() {
...
-            var administrativeAgreements = extractAdministrativeAgreements(bragePublicationRepresentation.publication());
+        if (!hasAdministrativeAgreement(cristinPublication) && hasAdministrativeAgreement(
+            bragePublicationRepresentation.publication())) {
+            var administrativeAgreements = extractAdministrativeAgreements(
+                bragePublicationRepresentation.publication());
             cristinPublication.getAssociatedArtifacts().addAll(administrativeAgreements);
             return cristinPublication.getAssociatedArtifacts();
         }
-        if (shouldUseBrageArtifacts()) {
+        if (shouldOverWriteWithBrageArtifacts()) {
```
Comment (anetteOlli) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java:186]: Not sure myself about this.
Do we want to replace everything when one of the files are better than the one present?

File: brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java [hunk 4]
```diff
@@ -76,13 +80,13 @@ public CristinImportPublicationMerger(Publication existingPublication, Publicati
     public Publication mergePublications()
         throws InvalidIsbnException, InvalidUnconfirmedSeriesException, InvalidIssnException {
         preMergeValidation();
-        var publicationForUpdating = existingPublication.copy()
-                                         .withAdditionalIdentifiers(mergeAdditionalIdentifiers())
-                                         .withSubjects(determineSubject())
-                                         .withRightsHolder(determineRightsHolder())
-                                         .withEntityDescription(determineEntityDescription())
-                                         .build();
-        return fillNewPublicationWithMetadataFromBrage(publicationForUpdating);
```
Comment (anetteOlli) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java:85]: flyttet de få metodene fra fillNewPublicationWithMetadataFromBrage inn hit eller inn i determineEntityDescription metoden, slik at det ikke blir så komplisert å finne ut hvor ting hentes ut.

**Review by truhacevkir — APPROVED** [review]
Prat med Terje, kanskje vi ønsker å beholde alle administrativeAgreements, usikker.

File: brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java [hunk 5]
```diff
@@ -152,33 +161,75 @@ private Set<AdditionalIdentifier> mergeAdditionalIdentifiers() {
...
+    }
+
+    private List<PublishedFile> extractPublishedFiles(Publication publication) {
+        return publication.getAssociatedArtifacts()
+                   .stream()
+                   .filter(associatedArtifact -> associatedArtifact instanceof File)
+                   .map(associatedArtifact -> (File) associatedArtifact)
+                   .filter(file -> file instanceof PublishedFile)
+                   .map(file -> (PublishedFile) file)
+                   .toList();
```
Comment (truhacevkir) [inline brage-import/src/main/java/no/sikt/nva/brage/migration/merger/CristinImportPublicationMerger.java:212]: I think we can filter directly by PublishedFile:

```
return publication.getAssociatedArtifacts()
                   .stream()
                   .filter(PublishedFile.class::isInstance)
                   .map(PublishedFile.class::cast)
                   .toList();
```

**Review by truhacevkir — APPROVED** [review]

**Pull request merged**